external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2937,12 +2940,82 @@ fn module_call_impl(
     (call, return_)
 }
 
+/// If the constant that we are referencing has an `error` attribute, we need to change the type of
+/// the constant to a u64 since this will be compiled into a u64 error code.
+fn annotated_error_const(context: &mut Context, e: &mut T::Exp) {
+    if let sp!(
+        const_loc,
+        T::UnannotatedExp_::Constant(module_ident, constant_name)
+    ) = &mut e.exp
+    {
+        let ConstantInfo {
+            attributes,
+            defined_loc,
+            signature,
+        } = context.constant_info(module_ident, constant_name);
+        let has_error_annotation =
+            attributes.contains_key_(&known_attributes::ErrorAttribute.into());
+        let u64_type = matches!(
+            signature.value.builtin_name(),
+            Some(sp!(_, N::BuiltinTypeName_::U64))
+        );
+        let ty_loc = e.ty.loc;
+        let const_def_loc = *defined_loc;
+
+        if has_error_annotation
+            && context.env.check_feature(
+                context.current_package(),
+                FeatureGate::CleverAssertions,
+                *const_loc,
+            )
+        {
+            e.ty = N::Type_::builtin(ty_loc, sp(ty_loc, N::BuiltinTypeName_::U64), vec![]);
+        }
+
+        // Add help messages
+        if !has_error_annotation
+                && !u64_type
+                    // If they're trying to use a non-u64 constant as an error code and in legacy,
+                    // nudge them that this is available in Move 2024.
+                    && context.env.check_feature(
+                        context.current_package(),
+                        FeatureGate::CleverAssertions,
+                        *const_loc,
+                        )
+        {
+            // If in Move 2024, tell them that they probably meant to add an error attribute on
+            // the constant.
+            let msg = format!(
+                "Invalid use of a non-u64 constant '{}' as error code.",
+                constant_name
+            );
+            let msg2 = format!(
+                "'{}' defined here with no '#[{}]' annotation",
+                constant_name,
+                known_attributes::ErrorAttribute
+            );
+            let mut err = diag!(
+                TypeSafety::InvalidCallTarget,
```

Comment:
Not sure this is the right error. Maybe we should just have a new one for bad `error`s?

Reply:
Sounds good to me. This was the best-closest one I could find, but adding a new one for this would make sense.